proc/proc_windows.go

```diff
@@ -0,0 +1,369 @@
+}
+
+func (dbp *Process) parseDebugLineInfo(exe *pe.File, wg *sync.WaitGroup) {
+	defer wg.Done()
+		
+	if sec := exe.Section(".debug_line"); sec != nil {
+		debugLine, err := sec.Data()
+		if err != nil && uint32(len(debugLine)) < sec.Size {
+			fmt.Println("could not get .debug_line section", err)
+			os.Exit(1)
+		}
+		if 0 < sec.VirtualSize && sec.VirtualSize < sec.Size {
+			debugLine = debugLine[:sec.VirtualSize]
+		}
+		dbp.lineInfo = line.Parse(debugLine)
+	} else {
+		fmt.Println("could not find .debug_line section in binary")
+		os.Exit(1)
+	}
+}
+
+func (dbp *Process) findExecutable(path string) (*pe.File, error) {
+	if path == "" {
+		// TODO: Find executable path from PID/handle on Windows:
+		// https://msdn.microsoft.com/en-us/library/aa366789(VS.85).aspx
+		return nil, fmt.Errorf("Not yet implemented")
+	}
+	f, err := os.OpenFile(path, 0, os.ModePerm)
+	if err != nil {
+		return nil, err
+	}
+	peFile, err := pe.NewFile(f)
+	if err != nil {
+		return nil, err
+	}
+	data, err := peFile.DWARF()
+	if err != nil {
+		return nil, err
+	}
+	dbp.dwarf = data
+	return peFile, nil
+}
+
+func (dbp *Process) waitForDebugEvent() (threadID, exitCode int, err error) {
+	var debugEvent C.DEBUG_EVENT
+	for {
+		if C.WaitForDebugEvent(&debugEvent, C.INFINITE) == C.WINBOOL(0) { 
+			return 0, 0, fmt.Errorf("Could not WaitForDebugEvent")
+		}
+		unionPtr := unsafe.Pointer(&debugEvent.u[0])
+		switch debugEvent.dwDebugEventCode {
+		case C.CREATE_PROCESS_DEBUG_EVENT:
+			debugInfo := (*C.CREATE_PROCESS_DEBUG_INFO)(unionPtr)
+			dbp.os.hProcess = syscall.Handle(unsafe.Pointer(debugInfo.hProcess))
+			_, err = dbp.addThread(syscall.Handle(unsafe.Pointer(debugInfo.hThread)), int(debugEvent.dwThreadId), false)
+			if err != nil {
+				return 0, 0, err
+			}
+			
+			C.ContinueDebugEvent(debugEvent.dwProcessId, debugEvent.dwThreadId, C.DBG_CONTINUE)
+			continue
+		case C.CREATE_THREAD_DEBUG_EVENT:
+			debugInfo := (*C.CREATE_THREAD_DEBUG_INFO)(unionPtr)
+
+			_, err = dbp.addThread(syscall.Handle(unsafe.Pointer(debugInfo.hThread)), int(debugEvent.dwThreadId), false)
+			if err != nil {
+				return 0, 0, err
+			}
+
+			C.ContinueDebugEvent(debugEvent.dwProcessId, debugEvent.dwThreadId, C.DBG_CONTINUE)
+			continue
+		case C.LOAD_DLL_DEBUG_EVENT, C.UNLOAD_DLL_DEBUG_EVENT, C.EXIT_THREAD_DEBUG_EVENT, C.OUTPUT_DEBUG_STRING_EVENT, C.RIP_EVENT:
+			C.ContinueDebugEvent(debugEvent.dwProcessId, debugEvent.dwThreadId, C.DBG_CONTINUE)
+			continue
+		case C.EXCEPTION_DEBUG_EVENT:
+			dbp.os.breakThread = int(debugEvent.dwThreadId)
+			return int(debugEvent.dwThreadId), 0, nil
+		case C.EXIT_PROCESS_DEBUG_EVENT:
+			debugInfo := (*C.EXIT_PROCESS_DEBUG_INFO)(unionPtr)
+			return 0, int(debugInfo.dwExitCode), nil
+		default:
+			return 0, 0, fmt.Errorf("Unknown event code: ", debugEvent.dwDebugEventCode)	
+		}
+	}
+}
+
+func (dbp *Process) trapWait(pid int) (*Thread, error) {	
+	for {
+		var err error
+		var tid int
+		var exitCode int
+		dbp.execPtraceFunc(func() {
```

Comment:
Is there a restriction in Windows that these syscalls must originate from the same thread?